### PR TITLE
Revise `cardano-balance-tx` module hierarchy.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3185

## Description

This PR makes a number of revisions to the `cardano-balance-tx` module hierarchy. We:
1. Remove module `Cardano.Tx.Balance`.
    - Justification: this was empty and unused.
1. Rename module namespace `Cardano.Wallet.Write` to `Cardano.Write`.
    - Justification: these modules are no longer specific to the wallet software.
1. Rename module namespace `Cardano.Tx.Balance.Internal.CoinSelection` to `Internal.Cardano.Write.Tx.Balance.CoinSelection`.
    - Justification: giving all internal modules a common prefix makes it trivial to search for all usages of internal modules.
1. Move module `Cardano.Write.Tx.Balance.TokenBundleSizeSpec` to `cardano-balance-tx` library.
    - Justification: the `TokenBundleSize` module is already in `cardano-balance-tx`, so it makes sense to colocate its test suite in the same library.

## Future work

- Continue to clarify and minimise the public interface of the `cardano-balance-tx` library.
- Move internal functionality into `Internal.Cardano.Write` module hierarchy (so all internal modules share a common prefix).
- Move `Cardano.Write.TxSpec` into `cardano-balance-tx`.
  - This is currently blocked by a dependency on `Cardano.Api.Gen`, which is currently defined in `cardano-wallet`.
  - Therefore, we'd first need to move `Cardano.Api.Gen` into a separate library on which both `cardano-wallet` and `cardano-balance-tx` can depend. (See #4143.)

### CODE REVIEW

File: lib/balance-tx/lib/Cardano/Tx/Balance.hs [hunk 1]
```diff
@@ -1,7 +0,0 @@
--- |
--- Copyright: © 2022 IOHK
--- License: Apache-2.0
---
-module Cardano.Tx.Balance
-    ()
-    where
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/Cardano/Tx/Balance.hs:7]: I think @Unisay will be happy that this module has gone 😄

File: lib/balance-tx/cardano-balance-tx.cabal [hunk 2]
```diff
@@ -67,10 +65,12 @@ library
     , generics-sop
     , int-cast
     , lens
+    , MonadRandom
     , monoid-subclasses
     , ouroboros-consensus
     , ouroboros-consensus-cardano
     , pretty-simple
+    , QuickCheck
```
Comment (Anviking) [inline lib/balance-tx/cardano-balance-tx.cabal:73]: We seem to have different methods of sorting 😄

File: lib/balance-tx/lib/Cardano/Write/Tx/Balance.hs [hunk 3]
```diff
@@ -23,7 +23,7 @@
 {-# OPTIONS_GHC -fno-warn-ambiguous-fields #-}
 #endif
 
-module Cardano.Wallet.Write.Tx.Balance
+module Cardano.Write.Tx.Balance
```
Comment (Anviking) [inline lib/balance-tx/lib/Cardano/Write/Tx/Balance.hs:26]: 🙌

File: lib/balance-tx/cardano-balance-tx.cabal [hunk 4]
```diff
@@ -67,30 +65,31 @@ library
...
+    Cardano.Write.Tx.Balance
+    Cardano.Write.Tx.Balance.TokenBundleSize
+    Cardano.Write.Tx.Gen
+    Cardano.Write.Tx.Redeemers
+    Cardano.Write.Tx.Sign
+    Cardano.Write.Tx.SizeEstimation
+    Cardano.Write.Tx.TimeTranslation
+    Cardano.Write.UTxOAssumptions
+    Internal.Cardano.Write.Tx.Balance.CoinSelection
+    Internal.Cardano.Write.Tx.Balance.CoinSelection.Gen
```
Comment (Anviking) [inline lib/balance-tx/cardano-balance-tx.cabal:92]: I'm guessing this is the beginning of a pattern? I'm not fully convinced, but happy to see where it takes us and/or to discuss further when convenient.

**Review by Anviking — APPROVED** [review]
Thank you! 🙌

File: lib/balance-tx/cardano-balance-tx.cabal (hunk 2 above)
Comment (Anviking) [reply lib/balance-tx/cardano-balance-tx.cabal:73]: vim's `:sort i` seems to do the trick

File: lib/balance-tx/cardano-balance-tx.cabal (hunk 4 above)
Comment (Anviking) [reply lib/balance-tx/cardano-balance-tx.cabal:92]: > Move internal functionality into Internal.Cardano.Write module hierarchy (so all internal modules share a common prefix).

Could we discuss this first?

File: lib/balance-tx/cardano-balance-tx.cabal (hunk 2 above)
Comment (jonathanknowles) [reply lib/balance-tx/cardano-balance-tx.cabal:73]: > We seem to have different methods of sorting 😄

I sorted with `sort i` (which is case insensitive)

File: lib/balance-tx/lib/Cardano/Tx/Balance.hs (hunk 1 above)
Comment (Unisay) [reply lib/balance-tx/lib/Cardano/Tx/Balance.hs:7]: Bye bye, you've been a good module!

File: lib/balance-tx/cardano-balance-tx.cabal (hunk 4 above)
Comment (jonathanknowles) [reply lib/balance-tx/cardano-balance-tx.cabal:92]: > Could we discuss this first?

Sure. Happy to discuss all options!

**Pull request merged**